### PR TITLE
Use Eclipse Collections' primitive HashSet IntHashSet

### PR DESCRIPTION
**Pull request by mdoube**
`ConnectedComponents` leans heavily on `HashSet<Integer>` but it is bloated and slow. [Eclipse Collections](https://github.com/eclipse/eclipse-collections) has a faster primitive `IntHashSet`. This PR swaps out `HashSet<Integer>` for `IntHashSet`. Performance improvement for larger images is on the order of 20%

```
//With IntHashSet
Particle labelling finished in 56 ms
Particle labelling finished in 53 ms
Particle labelling finished in 58 ms
Particle labelling finished in 43 ms
Particle labelling finished in 51 ms
Particle labelling finished in 222 ms
Particle labelling finished in 225 ms
Particle labelling finished in 258 ms
Particle labelling finished in 198 ms
Particle labelling finished in 253 ms
Particle labelling finished in 1156 ms
Particle labelling finished in 833 ms
Particle labelling finished in 1006 ms
Particle labelling finished in 758 ms
Particle labelling finished in 931 ms
Particle labelling finished in 10807 ms
Particle labelling finished in 11290 ms
Particle labelling finished in 9787 ms
Particle labelling finished in 10237 ms
Particle labelling finished in 9911 ms
Particle labelling finished in 83828 ms
Particle labelling finished in 75762 ms
Particle labelling finished in 75234 ms
Particle labelling finished in 64712 ms
Particle labelling finished in 67332 ms
//bonej-7.0.1
Particle labelling finished in 53 ms
Particle labelling finished in 48 ms
Particle labelling finished in 68 ms
Particle labelling finished in 55 ms
Particle labelling finished in 70 ms
Particle labelling finished in 237 ms
Particle labelling finished in 214 ms
Particle labelling finished in 239 ms
Particle labelling finished in 208 ms
Particle labelling finished in 182 ms
Particle labelling finished in 1021 ms
Particle labelling finished in 1077 ms
Particle labelling finished in 1359 ms
Particle labelling finished in 1346 ms
Particle labelling finished in 1257 ms
Particle labelling finished in 14164 ms
Particle labelling finished in 15563 ms
Particle labelling finished in 11314 ms
Particle labelling finished in 12564 ms
Particle labelling finished in 11975 ms
Particle labelling finished in 90115 ms
Particle labelling finished in 101046 ms
Particle labelling finished in 84493 ms
Particle labelling finished in 93437 ms
Particle labelling finished in 84424 ms
```

### CODE REVIEW

**Comment by rimadoma** [comment]
Looks good. @mdoube There was some kind of hiccup on the Travis build for 94a4761, but when I ran it again it passed. Seems to be passing now too.

File: Legacy/bonej/pom.xml [hunk 1]
```diff
@@ -1,170 +1,160 @@
...
-	<parent>
-		<groupId>org.bonej</groupId>
-		<artifactId>pom-bonej</artifactId>
-		<version>7.0.2-SNAPSHOT</version>
-		<relativePath>../..</relativePath>
-	</parent>
+    <parent>
+        <groupId>org.bonej</groupId>
+        <artifactId>pom-bonej</artifactId>
+        <version>7.0.1-SNAPSHOT</version>
```
Comment (mdoube) [inline Legacy/bonej/pom.xml:8]: Are you sure about changing this? Pretty sure we are currently on 7.0.2-SNAPSHOT

File: Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java [hunk 2]
```diff
@@ -648,11 +646,8 @@ private static void applyLUT(final int[][] particleLabels, final int[][] lut, fi
 	 */
 	private static void addNeighboursToMap(final List<IntHashSet> map, final int[] nbh, final int centre,
 			final int IDoffset) {
-		final int l = nbh.length;
 		int lastNonZero = -1;
-		for (int i = 0; i < l; i++) {
-			final int val = nbh[i];
-
+		for (final int val : nbh) {
```
Comment (mdoube) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java:650]: This is a really hot bit of code and minor changes can have performance implications. The longhand standard for loop checking against a `final int` is the fastest usually. Please only change if it's been benchmarked to make a speed improvement.

File: Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java [hunk 3]
```diff
@@ -673,11 +668,9 @@ private static void addNeighboursToMap(final List<IntHashSet> map, final int[] n
 	 * @param centre current pixel's map index (label - IDoffset)
 	 */
 	private static void addChunkNeighboursToMap(final List<IntHashSet> map, final int[] nbh, final int centre) {
-		final int l = nbh.length;
 		final IntHashSet set = map.get(centre);
 		int lastNonZero = -1;
-		for (int i = 0; i < l; i++) {
-			final int val = nbh[i];
+		for (final int val : nbh) {
```
Comment (mdoube) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java:673]: Same as above, no change unless faster.

File: Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java [hunk 4]
```diff
@@ -942,14 +935,11 @@ private static boolean withinBounds(final int m, final int n, final int o, final
...
 	private static int getMinTag(final int[] neighbourhood, final int ID) {
-		final int l = neighbourhood.length;
 		int minTag = ID;
-		for (int i = 0; i < l; i++) {
-			final int tagv = neighbourhood[i];
-			if (tagv == 0)
-				continue;
-			if (tagv < minTag)
+		for (final int tagv : neighbourhood) {
+			if (tagv != 0 && tagv < minTag) {
```
Comment (mdoube) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java:940]: Same again. May be an advantage in making `tagv` `final`. But I don't believe it until I see the numbers.

File: Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java [hunk 5]
```diff
@@ -979,8 +969,7 @@ private static void expandMap(final List<IntHashSet> map, final int ID,
 	 * @return number of particles in the image
 	 */
 	public int getNParticles() {
-		final int n = nParticles;
-		return n;
+		return nParticles;
 	}
```
Comment (mdoube) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java:973]: Is it safe to return a field like this?

File: Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java [hunk 6]
```diff
@@ -123,7 +123,7 @@ public ConnectedComponents() {
 		final int[][] lut = generateLut(chunkMaps, chunkIDOffsets);
 		
 		// rewrite the pixel values using the LUT
-		applyLUT(particleLabels, lut, chunkIDOffsets, startSlices, w, h, nSlices);
+		applyLUT(particleLabels, lut, chunkIDOffsets, startSlices, nSlices);
```
Comment (mdoube) [inline Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java:126]: Good catch.

File: Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java (hunk 5 above)
Comment (rimadoma) [reply Legacy/bonej/src/main/java/org/bonej/plugins/ConnectedComponents.java]: Yes. It's a primitive field, so value is returned.

**Comment by mdoube** [comment]
Benchmarking `for` loop style change:

```
//old-style for (int i = 0; i < max; i++)
Particle labelling finished in 51 ms
Particle labelling finished in 47 ms
Particle labelling finished in 66 ms
Particle labelling finished in 50 ms
Particle labelling finished in 62 ms
Particle labelling finished in 263 ms
Particle labelling finished in 242 ms
Particle labelling finished in 222 ms
Particle labelling finished in 184 ms
Particle labelling finished in 225 ms
Particle labelling finished in 988 ms
Particle labelling finished in 1070 ms
Particle labelling finished in 862 ms
Particle labelling finished in 899 ms
Particle labelling finished in 934 ms
Particle labelling finished in 11788 ms
Particle labelling finished in 11796 ms
Particle labelling finished in 10078 ms
Particle labelling finished in 10078 ms
Particle labelling finished in 9192 ms
//for (final int val : array) style loops
Particle labelling finished in 53 ms
Particle labelling finished in 60 ms
Particle labelling finished in 52 ms
Particle labelling finished in 48 ms
Particle labelling finished in 51 ms
Particle labelling finished in 179 ms
Particle labelling finished in 182 ms
Particle labelling finished in 216 ms
Particle labelling finished in 195 ms
Particle labelling finished in 216 ms
Particle labelling finished in 1026 ms
Particle labelling finished in 1052 ms
Particle labelling finished in 974 ms
Particle labelling finished in 973 ms
Particle labelling finished in 822 ms
Particle labelling finished in 10627 ms
Particle labelling finished in 10757 ms
Particle labelling finished in 9247 ms
Particle labelling finished in 9230 ms
Particle labelling finished in 9243 ms
```

**Pull request merged**